File: genetic_classifier.py
import random
import numpy as np
from sklearn.metrics import accuracy_score


class Rule:
    def __init__(self, n_features, feature_ranges):
        self.label = random.choice([0, 1])
        self.conditions = []
        for _ in range(random.randint(1, 3)):
            f = random.randint(0, n_features - 1)
            lo, hi = feature_ranges[f]
            t = random.uniform(lo, hi)
            op = random.choice(['<', '>'])
            self.conditions.append((f, op, t))

    def matches(self, x):
        for f, op, t in self.conditions:
            if op == '<' and x[f] >= t:
                return False
            if op == '>' and x[f] <= t:
                return False
        return True

    def to_dict(self, feature_names):
        return {
            'bedingungen': [
                {
                    'merkmal': feature_names[f],
                    'operator': op,
                    'schwellwert': round(float(t), 4),
                }
                for f, op, t in self.conditions
            ],
            'klasse': int(self.label),
            'klasse_text': 'Steigt ↑' if self.label == 1 else 'Fällt ↓',
        }


class RuleSet:
    def __init__(self, rules):
        self.rules = rules

    def predict_one(self, x):
        for rule in self.rules:
            if rule.matches(x):
                return rule.label
        return 0  # Standardklasse: Kurs fällt

    def predict(self, X):
        return np.array([self.predict_one(x) for x in X])


class GeneticRuleClassifier:
    def __init__(self, population_size=50, generations=30,
                 n_survivors=10, complexity_penalty=0.01):
        self.population_size = population_size
        self.generations = generations
        self.n_survivors = n_survivors
        self.complexity_penalty = complexity_penalty
        self.best_model = None
        self.feature_names = None
        self.history = []
        self._n_features = None
        self._feature_ranges = None

    def _new_ruleset(self):
        n = random.randint(1, 5)
        return RuleSet([Rule(self._n_features, self._feature_ranges) for _ in range(n)])

    def _fitness(self, ruleset, X, y):
        preds = ruleset.predict(X)
        acc = accuracy_score(y, preds)
        complexity = sum(len(r.conditions) for r in ruleset.rules)
        return acc - self.complexity_penalty * complexity

    def _mutate(self, ind):
        rules = ind.rules[:]
        if random.random() < 0.5 and len(rules) > 1:
            del rules[random.randint(0, len(rules) - 1)]
        else:
            rules.append(Rule(self._n_features, self._feature_ranges))
        return RuleSet(rules)

    def _crossover(self, a, b):
        c1 = random.randint(0, len(a.rules))
        c2 = random.randint(0, len(b.rules))
        rules = a.rules[:c1] + b.rules[c2:]
        if not rules:
            rules = [Rule(self._n_features, self._feature_ranges)]
        return RuleSet(rules)

    def fit(self, X, y, feature_names=None, progress_callback=None):
        self._n_features = X.shape[1]
        self._feature_ranges = [
            (float(X[:, i].min()), float(X[:, i].max()))
            for i in range(self._n_features)
        ]
        self.feature_names = feature_names or [f'Merkmal_{i}' for i in range(self._n_features)]

        population = [self._new_ruleset() for _ in range(self.population_size)]
        self.history = []
        best_ever, best_score_ever = None, -np.inf

        for gen in range(self.generations):
            scored = sorted(
                [(ind, self._fitness(ind, X, y)) for ind in population],
                key=lambda x: x[1], reverse=True,
            )
            best, best_score = scored[0]
            self.history.append(round(best_score, 4))

            if best_score > best_score_ever:
                best_score_ever = best_score
                best_ever = best

            if progress_callback:
                progress_callback(gen + 1, self.generations, best_score)

            survivors = [ind for ind, _ in scored[:self.n_survivors]]
            new_pop = survivors[:]
            while len(new_pop) < self.population_size:
                if random.random() < 0.5:
                    new_pop.append(self._mutate(random.choice(survivors)))
                else:
                    p1, p2 = random.sample(survivors, 2)
                    new_pop.append(self._crossover(p1, p2))
            population = new_pop

        self.best_model = best_ever
        return self

    def predict(self, X):
        return self.best_model.predict(X)

    def get_rules(self):
        return [r.to_dict(self.feature_names) for r in self.best_model.rules]

File: xampp/lib/GeneticClassifier.php
<?php
declare(strict_types=1);

class Rule
{
    public array $conditions;
    public int   $label;

    public function __construct(int $nFeatures, array $featureRanges)
    {
        $this->label      = rand(0, 1);
        $this->conditions = [];
        $nConds = rand(1, 3);
        for ($i = 0; $i < $nConds; $i++) {
            $f        = rand(0, $nFeatures - 1);
            [$lo, $hi] = $featureRanges[$f];
            $t        = $lo + (mt_rand() / mt_getrandmax()) * ($hi - $lo);
            $op       = rand(0, 1) ? '<' : '>';
            $this->conditions[] = [$f, $op, $t];
        }
    }

    public function matches(array $x): bool
    {
        foreach ($this->conditions as [$f, $op, $t]) {
            if ($op === '<' && $x[$f] >= $t) return false;
            if ($op === '>' && $x[$f] <= $t) return false;
        }
        return true;
    }

    public function toArray(array $featureNames): array
    {
        return [
            'bedingungen' => array_map(fn($c) => [
                'merkmal'     => $featureNames[$c[0]],
                'operator'    => $c[1],
                'schwellwert' => round($c[2], 4),
            ], $this->conditions),
            'klasse'      => $this->label,
            'klasse_text' => $this->label === 1 ? 'Steigt ↑' : 'Fällt ↓',
        ];
    }
}

class RuleSet
{
    /** @var Rule[] */
    public array $rules = [];

    public function predictOne(array $x): int
    {
        foreach ($this->rules as $rule) {
            if ($rule->matches($x)) return $rule->label;
        }
        return 0; // Standardklasse
    }

    public function predict(array $X): array
    {
        return array_map(fn($x) => $this->predictOne($x), $X);
    }
}

class GeneticRuleClassifier
{
    private int   $populationSize;
    private int   $generations;
    private int   $nSurvivors;
    private float $complexityPenalty;
    private int   $nFeatures    = 0;
    private array $featureRanges = [];

    public ?RuleSet $bestModel    = null;
    public array    $featureNames = [];
    public array    $history      = [];

    public function __construct(
        int   $populationSize    = 50,
        int   $generations       = 30,
        int   $nSurvivors        = 10,
        float $complexityPenalty = 0.01
    ) {
        $this->populationSize    = $populationSize;
        $this->generations       = $generations;
        $this->nSurvivors        = $nSurvivors;
        $this->complexityPenalty = $complexityPenalty;
    }

    private function newRuleSet(): RuleSet
    {
        $rs = new RuleSet();
        $n  = rand(1, 5);
        for ($i = 0; $i < $n; $i++) {
            $rs->rules[] = new Rule($this->nFeatures, $this->featureRanges);
        }
        return $rs;
    }

    private function fitness(RuleSet $rs, array $X, array $y): float
    {
        $preds   = $rs->predict($X);
        $n       = count($y);
        $correct = 0;
        for ($i = 0; $i < $n; $i++) {
            if ($preds[$i] === $y[$i]) $correct++;
        }
        $acc        = $n > 0 ? $correct / $n : 0.0;
        $complexity = array_sum(array_map(fn($r) => count($r->conditions), $rs->rules));
        return $acc - $this->complexityPenalty * $complexity;
    }

    private function mutate(RuleSet $ind): RuleSet
    {
        $rs        = new RuleSet();
        $rs->rules = $ind->rules;
        if ((mt_rand() / mt_getrandmax()) < 0.5 && count($rs->rules) > 1) {
            array_splice($rs->rules, rand(0, count($rs->rules) - 1), 1);
        } else {
            $rs->rules[] = new Rule($this->nFeatures, $this->featureRanges);
        }
        return $rs;
    }

    private function crossover(RuleSet $a, RuleSet $b): RuleSet
    {
        $rs        = new RuleSet();
        $c1        = rand(0, count($a->rules));
        $c2        = rand(0, count($b->rules));
        $rs->rules = array_merge(
            array_slice($a->rules, 0, $c1),
            array_slice($b->rules, $c2)
        );
        if (empty($rs->rules)) {
            $rs->rules[] = new Rule($this->nFeatures, $this->featureRanges);
        }
        return $rs;
    }

    public function fit(array $X, array $y, array $featureNames = [], ?callable $cb = null): void
    {
        $this->nFeatures   = count($X[0]);
        $this->featureNames = $featureNames ?: array_map(fn($i) => "Merkmal_$i", range(0, $this->nFeatures - 1));

        $this->featureRanges = [];
        for ($i = 0; $i < $this->nFeatures; $i++) {
            $col = array_column($X, $i);
            $this->featureRanges[] = [min($col), max($col)];
        }

        $population    = [];
        for ($i = 0; $i < $this->populationSize; $i++) {
            $population[] = $this->newRuleSet();
        }

        $this->history = [];
        $bestEver      = null;
        $bestEverScore = -PHP_INT_MAX;

        for ($gen = 0; $gen < $this->generations; $gen++) {
            $scored = array_map(fn($ind) => [$ind, $this->fitness($ind, $X, $y)], $population);
            usort($scored, fn($a, $b) => $b[1] <=> $a[1]);

            /** @var RuleSet $best */
            [$best, $bestScore] = $scored[0];
            $this->history[] = round((float) $bestScore, 4);

            if ($bestScore > $bestEverScore) {
                $bestEverScore = $bestScore;
                $bestEver      = $best;
            }

            if ($cb !== null) {
                $cb($gen + 1, $this->generations, (float) $bestScore);
            }

            $survivors = array_map(fn($s) => $s[0], array_slice($scored, 0, $this->nSurvivors));
            $newPop    = $survivors;

            while (count($newPop) < $this->populationSize) {
                if ((mt_rand() / mt_getrandmax()) < 0.5) {
                    $newPop[] = $this->mutate($survivors[array_rand($survivors)]);
                } else {
                    $keys = array_rand($survivors, 2);
                    $newPop[] = $this->crossover($survivors[$keys[0]], $survivors[$keys[1]]);
                }
            }

            $population = $newPop;
        }

        $this->bestModel = $bestEver;
    }

    public function predict(array $X): array
    {
        return $this->bestModel->predict($X);
    }

    public function getRules(): array
    {
        return array_map(fn($r) => $r->toArray($this->featureNames), $this->bestModel->rules);
    }
}
